saml2/saml2-service-provider/core/src/main/java/org/springframework/security/saml2/provider/service/web/authentication/logout/OpenSamlLogoutRequestResolver.java

```diff
@@ -0,0 +1,243 @@
+import org.springframework.security.saml2.provider.service.authentication.Saml2Authentication;
+import org.springframework.security.saml2.provider.service.authentication.logout.Saml2LogoutRequest;
+import org.springframework.security.saml2.provider.service.registration.RelyingPartyRegistration;
+import org.springframework.security.saml2.provider.service.registration.Saml2MessageBinding;
+import org.springframework.security.saml2.provider.service.web.RelyingPartyRegistrationResolver;
+import org.springframework.util.Assert;
+
+/**
+ * A {@link Saml2LogoutRequestResolver} for resolving SAML 2.0 Logout Requests with
+ * OpenSAML
+ *
+ * Note that there are {@link Saml2LogoutRequestResolver} implements that are targeted for
+ * OpenSAML 3 and OpenSAML 4 via {@code OpenSaml3LogoutRequestResolver} and
+ * {@code OpenSaml4LogoutRequestResolver}
+ *
+ * @author Josh Cummings
+ * @since 5.5
+ */
+public final class OpenSamlLogoutRequestResolver implements Saml2LogoutRequestResolver {
+
+	private final RelyingPartyRegistrationResolver relyingPartyRegistrationResolver;
+
+	/**
+	 * Construct a {@link OpenSamlLogoutRequestResolver} using the provided parameters
+	 * @param relyingPartyRegistrationResolver the
+	 * {@link RelyingPartyRegistrationResolver} for selecting the
+	 * {@link RelyingPartyRegistration}
+	 */
+	public OpenSamlLogoutRequestResolver(RelyingPartyRegistrationResolver relyingPartyRegistrationResolver) {
+		this.relyingPartyRegistrationResolver = relyingPartyRegistrationResolver;
+	}
+
+	/**
+	 * Prepare to create, sign, and serialize a SAML 2.0 Logout Request.
+	 *
+	 * By default, includes a {@code NameID} based on the {@link Authentication} instance
+	 * as well as the {@code Destination} and {@code Issuer} based on the
+	 * {@link RelyingPartyRegistration} derived from the {@link Authentication}.
+	 *
+	 * The {@link Authentication} must be of type {@link Saml2Authentication} in order to
+	 * look up the {@link RelyingPartyRegistration} that holds the signing key.
+	 * @param request the HTTP request
+	 * @param authentication the current principal details
+	 * @return a builder, useful for overriding any aspects of the SAML 2.0 Logout Request
+	 * that the resolver supplied
+	 */
+	@Override
+	public OpenSamlLogoutRequestBuilder resolveLogoutRequest(HttpServletRequest request,
+			Authentication authentication) {
+		Assert.notNull(authentication, "Failed to lookup logged-in user for formulating LogoutRequest");
+		RelyingPartyRegistration registration = this.relyingPartyRegistrationResolver.resolve(request,
+				getRegistrationId(authentication));
+		Assert.notNull(registration, "Failed to lookup RelyingPartyRegistration for formulating LogoutRequest");
+		return new OpenSamlLogoutRequestBuilder(registration)
+				.destination(registration.getAssertingPartyDetails().getSingleLogoutServiceLocation())
+				.issuer(registration.getEntityId()).name(authentication.getName());
```

Comment:
It cannot be. Because OpenSAML 3 and 4 have incompatible types for IssueInstant. So, this is supplied in the `OpenSamlXLogoutRequestResolver` implementations.

Perhaps this class should be package private to make that clearer.

Reply:
I addressed this by making the `OpenSamlLogoutRequestResolver`'s constructor package private. I believe this will clarify that applications should use either `OpenSaml3LogoutRequestResolver` or `OpenSaml4LogoutRequestResolver`. Please feel free to reopen if you feel there's more to discuss.